### ISSUE TITLE
provider is not evaluating expression for 'application/x-thrift' protocol in 4.2.2 version

### ISSUE THREAD

**Issue by keeping-it-up** [pact-foundation/pact-jvm#1330]
With 4.2.2 version of pact, noticed that provider is not evaluating expression for 'application/x-thrift' protocol.

To provide background info, we pushed a PR as part of issue https://github.com/pact-foundation/pact-jvm/issues/1298 which allows to use json matcher for application/x-thrift protocol. In version 4.1.17 this change was working fine and expression were evaluating properly - getting value from state method.

In another issue https://github.com/pact-foundation/pact-jvm/issues/1318, array matching issue related to arrayContaining() was fixed in version 4.2.2.
 
However after upgrading pact version to 4.2.2, we have noticed that expressions are not getting value from state method for 'application/x-thrift' protocol.
With application/json or application/x-thrift+json protocol, expressions are fine.

I can see that json matcher is being used for 'application/x-thrift' protocol so code for issue 1298 is being merged into 4.2.2 (i can also verify this in git 4.2.x branch).

In terms of debug, I can see that invokeStateChangeMethods() function of PactVerificationStateChangeExtension calls my state method which sets expression value into map, then in beforeTestExecution() function provider state is set to testContext:
`
testContext.executionContext = mapOf("providerState" to providerStateContext)`

after that when flow comes to context.verifyInteraction(); method of my provider test class, i can see request still having example parameter set by consumer instead of provider's state method parameter.

Logs also say that state method is invoked:

```
23 Mar 2021 12:26:50,485 DEBUG PactVerificationStateChangeExtension:33 [main] beforeEach for interaction 'test request'
23 Mar 2021 12:27:04,659  INFO PactVerificationStateChangeExtension:103 [main] Invoking state change method 'test request':SETUP

Verifying a pact between test-consumer and test-provider for tag test-tag

```

**Comment by uglyog** [comment]
With the change applied to fix #1314 you can now set the type for the content type with the `pact.content_type.override.application/x-thrift` system property set to `json`.

See https://github.com/pact-foundation/pact-jvm/blob/master/provider/junit5/src/test/groovy/au/com/dius/pact/provider/junit5/ThriftStateInjectedProviderTest.groovy#L38 for an example.

**Comment by keeping-it-up** [comment]
Thanks Ronald.
I have this property set in both my consumer and producer test. Its also using json body matcher because of this property, however doesn't replace the expression value from state method.

**Comment by uglyog** [comment]
@keeping-it-up can you have a look at the test I posted above, that works when I run it.

**Comment by keeping-it-up** [comment]
I have run the test your referenced and it works.
However I have added a new test which has content type header in json and uses expression in request body - this test doesn't work.
It doesn't replace id value to abc and hence request doesn't match.

If I remove header part from request section of json:

```
"headers": {
          "Content-Type": ["application/x-thrift"]
        }
```
then it starts to work.

Thanks.

**Comment by keeping-it-up** [comment]
hi @uglyog, we have put a fix for this issue and a test case which now works after the fix: https://github.com/pact-foundation/pact-jvm/pull/1340

can you please review and merge if you are happy? please also release a new version so we can unblock our story.
Thanks.

**Comment by uglyog** [comment]
PR has been merged

**Comment by keeping-it-up** [comment]
thanks @uglyog, can you let me know if you have done a release and release number please? thanks.

**Comment by uglyog** [comment]
4.2.5 has been released

**Comment by keeping-it-up** [comment]
Works now, cheers.
